src/point_one/fusion_engine/messages/data_version.cc

```diff
@@ -6,18 +6,17 @@ namespace point_one {
 namespace fusion_engine {
 namespace messages {
 
-std::ostream& operator<<(std::ostream& stream, const DataVersion& ver) {
-  if (ver.IsValid()) {
-    return stream << (int)ver.major_version << "." << ver.minor_version;
-  } else {
-    return stream << "<invalid>";
-  }
+p1_ostream& operator<<(p1_ostream& stream, const DataVersion& ver) {
+  return stream << ToString(ver);
 }
 
 std::string ToString(const DataVersion& ver) {
-  std::stringstream ss;
-  ss << ver;
-  return ss.str();
+  if (ver.IsValid()) {
+    return std::to_string(ver.major_version) + "." +
```

Comment:
I'd suggest switching this change back. Generally, if `ToString()` returns a `std::string` I find it's better to have `ToString()` call `<<` instead of the other way around. That way, if someone wants to stream out the object directly `s << foo` they can without allocating heap for a `std::string`.

The exception to that is if the `ToString()` returns `constexpr const char*` like for enums. In that case, `operator<<()` should call `ToString()`.

Reply:
I understand why the original is preferred, I switched it to this so this function could work regardless of `P1_HAVE_STD_OSTREAM` and to avoid needing to add `P1_HAVE_STD_OSTREAM` checks. I guess the "best" thing is to have both.